File: src/AllKthMarkov/java/src/KthMarkov.java
import java.util.Enumeration;
import java.util.Hashtable;
import java.util.Vector;
import java.util.Map;
public class KthMarkov{
    //Attributs
    private Hashtable<Vector<Document>,MarkovNode> nodes;

    //Constructeur
    public KthMarkov(){//DONE
        nodes=new Hashtable<Vector<Document>,MarkovNode>();
    }

    //Getter
    public MarkovNode getDocs(Vector<Document> docs){//DONE
        if(!nodes.containsKey(docs))
            nodes.put(docs, new MarkovNode(docs));
        return nodes.get(docs);
    }
    public Enumeration<MarkovNode> getNodes(){//DONE
        return nodes.elements();
    }

    //Methodes
    public void add(KthMarkov markov){//TODO
        Enumeration<MarkovNode> markovNodes = markov.getNodes();
        while(markovNodes.hasMoreElements()){
            MarkovNode mn = markovNodes.nextElement();
            MarkovNode pere = getDocs(mn.getMaChaineDoc());
            for(Map.Entry<MarkovNode, Integer> ent : mn.getChild().entrySet()){
                MarkovNode child = getDocs(ent.getKey().getMaChaineDoc());
                pere.renforcerChaine(child, ent.getValue());
            }
        }
    }
    public void renforcerChaine(Vector<Document> oldDocs, Document nextDoc){//DONE
        //Creation du vecteur nouveau document
        Vector<Document> newDocs= new Vector<Document>();
        Vector<Document> oldDocsCp= new Vector<Document>();
        for(Document doc:oldDocs){
            newDocs.add(doc);
            oldDocsCp.add(doc);
        }
        newDocs.remove(0);
        newDocs.add(nextDoc);

        //Insertion de noeuds si ces derniers n'existent pas
        MarkovNode pere = getDocs(oldDocsCp);
        MarkovNode child = getDocs(newDocs);

        //Renforcement de la chaine
        pere.renforcerChaine(child);
    }
    public String toSave(){//DONE
        String out="";

        //Nb noeud
        out+=nodes.size();
        out+=" ";

        //Nb Liens
        int nbLiens=0;
        for(MarkovNode m:nodes.values())
            nbLiens+=m.getNbChild();
        out+=nbLiens;
        out+="\n";

        //Affichage Noeuds
        for(Map.Entry<Vector<Document>, MarkovNode> node:nodes.entrySet()){
            for(Document doc:node.getKey())
                out += "\""+doc.getChemin()+"\" ";
            out += "["+node.getValue().getTotal()+"]\n";
        }

        //Affichage liens
        for(Map.Entry<Vector<Document>, MarkovNode> node : nodes.entrySet()){
            for(Map.Entry<MarkovNode, Integer> child : node.getValue().getChild().entrySet()){
                for(Document doc : node.getKey())
                    out += "\""+doc.getChemin()+"\" ";
                out += "-("+child.getValue()+")- ";
                for(Document doc : child.getKey().getMaChaineDoc())
                    out += "\""+doc.getChemin()+"\" ";
                out += "\n";
            }
        }
        return out;
    }
public String toUML(){//DONE
        String out="";

        for(Map.Entry<Vector<Document>, MarkovNode> node:nodes.entrySet()){
            for(Map.Entry<MarkovNode, Integer> child : node.getValue().getChild().entrySet()){
                out += "  <";

                for(Document doc:node.getKey())
                    out += doc.getChemin()+",";

                out += "> -> <";

                for(Document doc: child.getKey().getMaChaineDoc())
                    out += doc.getChemin()+",";

                out += "> [label=\"";
                out += ((float)child.getValue()/(float)node.getValue().getTotal());
                out += "\"]\n";
            }
        }
        return out;
    }
}

File: src/cpp/src/KthMarkov.cpp
#include "KthMarkov.hpp"

void KthMarkov::addNewNode(vector<int> node, int poidsNode){//DONE
    if(getPDocs(node) == nullptr){
        nodes[node]=MarkovNode(node);
        nodes[node].setTotal(poidsNode);
    }
}
MarkovNode* KthMarkov::getPDocs(vector<int> docs){//DONE
    if(nodes.find(docs) == nodes.end())
        return nullptr;
    return &nodes[docs];
}
void KthMarkov::renforcerChaine(vector<int> oldDocs, int nextDoc){//DONE
    //Calcul chaine nouveaux documents
    vector<int> newDocs(oldDocs);
    newDocs.erase(newDocs.begin());
    newDocs.push_back(nextDoc);

    //Recuperation adresse oldMarkovNode
    if(getPDocs(oldDocs) == nullptr)
        nodes[oldDocs]=MarkovNode(oldDocs);
    MarkovNode* oldMarkovNode = &(nodes[oldDocs]);

    //Recuperation adresse newMarkovNode
    if(getPDocs(newDocs) == nullptr)
        nodes[newDocs]=MarkovNode(newDocs);
    MarkovNode* newMarkovNode = &(nodes[newDocs]);

    //Renforcement chaine
    oldMarkovNode->renforcerChaine(newMarkovNode);
}
string KthMarkov::toSave(){//DONE
    //Affiochage nombre de noeuds
    string out=to_string(nodes.size());
    out+=" ";

    //Affichage nombre de liens
    int nbLiens=0;
    for(pair<vector<int>, MarkovNode> node: nodes)
        nbLiens+=node.second.getNbChild();
    out+=to_string(nbLiens);
    out+="\n";

    //Affichage noeuds
    for(pair<vector<int>, MarkovNode> node: nodes){
        for(int doc:node.first)
            out += to_string(doc)+" ";
        out += "["+to_string(node.second.getTotal())+"]";
        //out += to_string(node.second.getTotal());
        out += "\n";
    }

    //Affichage liens
    for(pair<vector<int>, MarkovNode> node: nodes){
        for(pair<MarkovNode*, int> child: node.second.getChild()){
            for(int doc:node.first)
                out+=to_string(doc)+" ";
            out+="-("+to_string(child.second)+")- ";
            //out+=to_string(child.second)+" ";
            for(int doc:child.first->getMaChaineDoc())
                out+=to_string(doc)+" ";
            out+="\n";
        }
    }

    return out;
}
string KthMarkov::toUML(){//DONE
    //Affiochage nombre de noeuds
    string out="";

    //Affichage liens
    for(pair<vector<int>, MarkovNode> node: nodes){
        for(pair<MarkovNode*, int> child: node.second.getChild()){
            out += "  <";

            for(int doc:node.first)
                out+=to_string(doc)+",";

            out += "> -> <";

            for(int doc:child.first->getMaChaineDoc())
                out+=to_string(doc)+",";

            out += "> [label=\""+to_string((float)child.second/(float)node.second.getTotal())+"\"]\n";
        }
    }

    return out;
}
